File: CousinsBST.java
// TC : O(N)
// SC: O(H)
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
// Resursive
class Solution {

    TreeNode xParent, yParent;
    int xLvl, yLvl;
    public boolean isCousins(TreeNode root, int x, int y) {
        // base case
        if (root == null) return false;

        dfs(null, x, y, root, 0);

        return (xLvl == yLvl && xParent != yParent);
    }

    private void dfs(TreeNode parent, int x, int y, TreeNode root, int level) {
        // base case
        if (root == null) return;
        // check if value matches x
        if(root.val == x) {
            xParent = parent;
            xLvl  = level;
        }

        // check if value matches y
        if (root.val == y) {
            yParent = parent;
            yLvl = level;
        }

        // recursive cases
        if (xParent ==null || yParent == null) {
            dfs(root, x, y, root.left, level+1);
            dfs(root, x, y, root.right, level+1);
        }


    }
}

File: cousins bst.py
# TC = O(n)
# SC = O(n)
# run on LC

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isCousins(self, root: Optional[TreeNode], x: int, y: int) -> bool:
        if (root == None):
            return False

        queue = collections.deque([(root, 0)])

        while (queue):
            ls = len(queue)
            px = py = -1

            for i in range(ls):
                node, parent = queue.popleft()

                if (node.val == x):
                    px = parent
                elif (node.val == y):
                    py = parent

                if (node.left != None):
                    queue.append((node.left, node.val))
                if (node.right != None):
                    queue.append((node.right, node.val))

            if (px != -1 and py != -1 and px != py):
                return True

        return False
